File: go/src/methods/loadstrike_runner/with_nats_server_url_method_reference.go
package loadstrike_runner

        import (
        	"time"

        	loadstrike "loadstrike.com/sdk/go"
        )

        const withNatsServerUrlRunnerKey = "runner_dummy_orders_reference"

type WithNatsServerUrlMethodReference struct{}

type withNatsServerUrlTempConfigPaths struct {
	ConfigPath string
	InfraPath  string
}

type withNatsServerUrlOrdersReportingSink struct{}

func newWithNatsServerUrlOrdersReportingSink() withNatsServerUrlOrdersReportingSink {
	return withNatsServerUrlOrdersReportingSink{}
}
func (withNatsServerUrlOrdersReportingSink) SinkName() string {
	return "orders-sample-sink"
}
func (withNatsServerUrlOrdersReportingSink) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersReportingSink) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersReportingSink) SaveRealtimeStats([]loadstrike.LoadStrikeScenarioStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersReportingSink) SaveRealtimeMetrics(loadstrike.LoadStrikeMetricStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersReportingSink) SaveRunResult(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersReportingSink) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersReportingSink) Dispose() {}

type withNatsServerUrlOrdersRuntimePolicy struct{}

func newWithNatsServerUrlOrdersRuntimePolicy() withNatsServerUrlOrdersRuntimePolicy {
	return withNatsServerUrlOrdersRuntimePolicy{}
}
func (withNatsServerUrlOrdersRuntimePolicy) ShouldRunScenario(string) loadstrike.LoadStrikeBoolTask {
	return loadstrike.TaskFromBool(true)
}
func (withNatsServerUrlOrdersRuntimePolicy) BeforeScenario(string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersRuntimePolicy) AfterScenario(string, loadstrike.LoadStrikeScenarioRuntime) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersRuntimePolicy) BeforeStep(string, string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersRuntimePolicy) AfterStep(string, string, loadstrike.LoadStrikeReply) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

type withNatsServerUrlOrdersWorkerPlugin struct{}

func newWithNatsServerUrlOrdersWorkerPlugin() withNatsServerUrlOrdersWorkerPlugin {
	return withNatsServerUrlOrdersWorkerPlugin{}
}
func (withNatsServerUrlOrdersWorkerPlugin) PluginName() string {
	return "orders-sample-plugin"
}
func (withNatsServerUrlOrdersWorkerPlugin) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersWorkerPlugin) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersWorkerPlugin) GetData(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikePluginData] {
	return loadstrike.TaskFromResult(loadstrike.CreatePluginData("orders-sample-plugin"))
}
func (withNatsServerUrlOrdersWorkerPlugin) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withNatsServerUrlOrdersWorkerPlugin) Dispose() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

func withNatsServerUrlPerformOrderGetReply() loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeResponse.Ok("200", int64(128), "ok", loadstrike.TimeSpan(3*time.Millisecond))
}

func withNatsServerUrlExecuteOrderGet(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeStep.Run("get-order", context, func(loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
		return withNatsServerUrlPerformOrderGetReply()
	})
}

func withNatsServerUrlExecuteOrderGetAsync(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikeReply] {
	return loadstrike.TaskFromResult(withNatsServerUrlExecuteOrderGet(context))
}

func withNatsServerUrlBaselineScenario(names ...string) loadstrike.LoadStrikeScenario {
	name := "orders.get-by-id"
	if len(names) > 0 && names[0] != "" {
		name = names[0]
	}
	return loadstrike.CreateScenario(name, withNatsServerUrlExecuteOrderGet).
		WithLoadSimulations(loadstrike.LoadStrikeSimulation.IterationsForConstant(1, 1)).
		WithoutWarmUp()
}

func withNatsServerUrlBaseRunner() loadstrike.LoadStrikeRunner {
	return loadstrike.Create().
		AddScenario(withNatsServerUrlBaselineScenario()).
		WithRunnerKey(withNatsServerUrlRunnerKey).
		WithTestSuite("orders-reference").
		WithTestName("orders-get-by-id").
		WithoutReports()
}

func withNatsServerUrlBaseContext() loadstrike.LoadStrikeContext {
	return withNatsServerUrlBaseRunner().BuildContext()
}

func withNatsServerUrlHttpSource() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-source",
		Mode:          "Produce",
		TrackingField: "header:X-Correlation-Id",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                   "https://orders.example.test/api/orders",
			Method:                "GET",
			TrackingPayloadSource: "Request",
			ResponseSource:        "ResponseBody",
		},
	}
}

func withNatsServerUrlHttpDestination() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-destination",
		Mode:          "Consume",
		TrackingField: "json:$.trackingId",
		GatherByField: "json:$.tenantId",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                      "https://orders.example.test/api/order-events",
			Method:                   "GET",
			ResponseSource:           "ResponseBody",
			ConsumeJSONArrayResponse: true,
			ConsumeArrayPath:         "$.items",
		},
	}
}

func withNatsServerUrlTrackingConfiguration() *loadstrike.TrackingConfigurationSpec {
	return &loadstrike.TrackingConfigurationSpec{
		Source:                      withNatsServerUrlHttpSource(),
		Destination:                 withNatsServerUrlHttpDestination(),
		RunMode:                     "GenerateAndCorrelate",
		CorrelationTimeoutSeconds:   30,
		TimeoutSweepIntervalSeconds: 1,
		TimeoutBatchSize:            200,
		TimeoutCountsAsFailure:      true,
		MetricPrefix:                "orders_tracking",
		ExecuteOriginalScenarioRun:  false,
	}
}

func withNatsServerUrlTrackedContext() loadstrike.LoadStrikeContext {
	return loadstrike.Create().
		AddScenario(withNatsServerUrlBaselineScenario("orders.tracked").WithCrossPlatformTracking(withNatsServerUrlTrackingConfiguration())).
		WithRunnerKey(withNatsServerUrlRunnerKey).
		WithoutReports().
		BuildContext()
}

func withNatsServerUrlBuildTrackingPayload() loadstrike.TrackingPayload {
	builder := loadstrike.TrackingPayloadBuilder{
		Headers:            map[string]string{"X-Correlation-Id": "ord-1"},
		ContentType:        "application/json",
		MessagePayloadType: "json",
	}
	builder.SetBody(map[string]any{"trackingId": "ord-1"})
	return builder.Build()
}

func withNatsServerUrlRunResult() loadstrike.LoadStrikeRunResult {
	var result loadstrike.LoadStrikeRunResult
	return result
}

func withNatsServerUrlScenarioStats() loadstrike.LoadStrikeScenarioStats {
	var stats loadstrike.LoadStrikeScenarioStats
	return stats
}

func withNatsServerUrlWriteTempConfigFiles() withNatsServerUrlTempConfigPaths {
	return withNatsServerUrlTempConfigPaths{
		ConfigPath: "method-reference.loadstrike.config.json",
		InfraPath:  "method-reference.loadstrike.infra.json",
	}
}

        // Apply the primary string value shown in the sample method reference.
func (reference WithNatsServerUrlMethodReference) ApplyPrimaryValueExample() any {
    return withNatsServerUrlBaseRunner().WithNatsServerURL("nats://localhost:4222")
}

// Show the same method with a second concrete value.
func (reference WithNatsServerUrlMethodReference) ApplyAlternateValueExample() any {
    return withNatsServerUrlBaseRunner().WithNatsServerURL("nats://demo-cluster:4222")
}

File: dotnet/src/LoadStrike.SampleReference/Methods/LoadStrikeRunner/WithNatsServerUrlMethodReference.cs
using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using LoadStrike;
    using LoadStrike.Contracts.Metrics;
    using LoadStrike.CrossPlatform;
    using LoadStrike.CrossPlatform.Http;
    using Microsoft.Extensions.Configuration;
    using Serilog;
    using Serilog.Events;

    namespace LoadStrike.SampleReference.Methods.OwnerLoadStrikeRunner;

    /// <summary>Configure the NATS endpoint used by clustered execution.</summary>
    public static class WithNatsServerUrlMethodReference
    {
        private const string RunnerKey = "runner_dummy_orders_reference";

private static Task<LoadStrikeReply<Dictionary<string, object>>> CreateTypedOrderReplyAsync()
{
    return Task.FromResult(
        LoadStrikeResponse.Ok(
            new Dictionary<string, object>(StringComparer.Ordinal)
            {
                ["orderId"] = "ORD-10001",
            },
            statusCode: "200",
            sizeBytes: 128,
            message: "ok",
            customLatencyMs: 3.2
        )
    );
}

private static LoadStrikeScenario baselineScenario(string name = "orders.get-by-id")
{
    return LoadStrikeScenario
        .Create(name, ExecuteOrderGetAsync)
        .WithLoadSimulations(LoadStrikeSimulation.IterationsForConstant(1, 1))
        .WithoutWarmUp();
}

private static async Task<LoadStrikeReply> ExecuteOrderGetAsync(LoadStrikeScenarioContext context)
{
    var reply = await LoadStrikeStep.Run("get-order", context, CreateTypedOrderReplyAsync);
    return reply.AsReply();
}

private static LoadStrikeRunner baseRunner()
{
    return LoadStrikeRunner
        .Create()
        .AddScenario(baselineScenario())
        .Configure(context => context.WithRunnerKey(RunnerKey))
        .WithTestSuite("orders-reference")
        .WithTestName("orders-get-by-id")
        .WithoutReports();
}

private static LoadStrikeContext baseContext() => baseRunner().BuildContext();

private static LoadStrikeContext trackedContext()
{
    return LoadStrikeRunner
        .Create()
        .AddScenario(
            CrossPlatformScenarioConfigurator.Configure(
                baselineScenario("orders.tracked"),
                trackingConfiguration()
            )
        )
        .Configure(context => context.WithRunnerKey(RunnerKey))
        .WithoutReports()
        .BuildContext();
}

private static HttpEndpointDefinition httpSource() =>
    new()
    {
        Name = "orders-http-source",
        Mode = TrafficEndpointMode.Produce,
        TrackingField = TrackingFieldSelector.Parse("header:X-Correlation-Id"),
        Url = "https://orders.example.test/api/orders",
        Method = "GET",
        TrackingPayloadSource = HttpTrackingPayloadSource.Request,
        ResponseSource = HttpResponseSource.ResponseBody,
    };

private static HttpEndpointDefinition httpDestination() =>
    new()
    {
        Name = "orders-http-destination",
        Mode = TrafficEndpointMode.Consume,
        TrackingField = TrackingFieldSelector.Parse("json:$.trackingId"),
        GatherByField = TrackingFieldSelector.Parse("json:$.tenantId"),
        Url = "https://orders.example.test/api/order-events",
        Method = "GET",
        ResponseSource = HttpResponseSource.ResponseBody,
        ConsumeJsonArrayResponse = true,
        ConsumeArrayPath = "$.items",
    };

private static CrossPlatformTrackingConfiguration trackingConfiguration()
{
    return new CrossPlatformTrackingConfiguration
    {
        Source = httpSource(),
        Destination = httpDestination(),
        RunMode = TrackingRunMode.GenerateAndCorrelate,
        CorrelationTimeout = TimeSpan.FromSeconds(30),
        TimeoutSweepInterval = TimeSpan.FromSeconds(1),
        TimeoutBatchSize = 200,
        TimeoutCountsAsFailure = true,
        MetricPrefix = "orders_tracking",
        ExecuteOriginalScenarioRun = false,
        CorrelationStore = CorrelationStoreConfiguration.InMemory(),
    };
}

private static TrackingPayload buildTrackingPayload()
{
    var builder = new TrackingPayloadBuilder();
    builder.SetBody("{\"trackingId\":\"ord-1\"}");
    return builder.Build();
}

private static LoadStrikeRunResult runResult() => default!;

private static LoadStrikeScenarioStats scenarioStats() => default!;

private static LoggerConfiguration buildLogger() => new();

private static TempConfigPaths writeTempConfigFiles()
{
    return new TempConfigPaths("method-reference.loadstrike.config.json", "method-reference.loadstrike.infra.json");
}

private readonly record struct TempConfigPaths(string ConfigPath, string InfraPath);

private sealed class OrdersReportingSink : LoadStrikeReportingSink
{
    public string SinkName => "orders-sample-sink";
    public Task Init(LoadStrikeBaseContext context, IConfiguration infraConfig) => Task.CompletedTask;
    public Task Start(LoadStrikeSessionStartInfo sessionInfo) => Task.CompletedTask;
    public Task SaveRealtimeStats(LoadStrikeScenarioStats[] stats) => Task.CompletedTask;
    public Task SaveRealtimeMetrics(LoadStrikeMetricStats metrics) => Task.CompletedTask;
    public Task SaveRunResult(LoadStrikeRunResult result) => Task.CompletedTask;
    public Task Stop() => Task.CompletedTask;
    public void Dispose() { }
}

private sealed class OrdersRuntimePolicy : ILoadStrikeRuntimePolicy
{
    public Task<bool> ShouldRunScenario(string scenarioName) => Task.FromResult(true);
    public Task BeforeScenario(string scenarioName) => Task.CompletedTask;
    public Task AfterScenario(string scenarioName, LoadStrikeScenarioRuntime stats) => Task.CompletedTask;
    public Task BeforeStep(string scenarioName, string stepName) => Task.CompletedTask;
    public Task AfterStep(string scenarioName, string stepName, LoadStrikeReply reply) => Task.CompletedTask;
}

private sealed class OrdersWorkerPlugin : ILoadStrikeWorkerPlugin
{
    public string PluginName => "orders-sample-plugin";
    public Task Init(LoadStrikeBaseContext context, IConfiguration infraConfig) => Task.CompletedTask;
    public Task Start(LoadStrikeSessionStartInfo sessionInfo) => Task.CompletedTask;
    public Task<LoadStrikePluginData> GetData(LoadStrikeRunResult result) => Task.FromResult(LoadStrikePluginData.Create(PluginName));
    public Task Stop() => Task.CompletedTask;
    public Task Dispose() => Task.CompletedTask;
}

        /// <summary>Apply the primary string value shown in the sample method reference.</summary>
public static object? ApplyPrimaryValueExample()
{
    return LoadStrikeRunner.WithNatsServerUrl(baseContext(), "nats://localhost:4222");
}

/// <summary>Show the same method with a second concrete value.</summary>
public static object? ApplyAlternateValueExample()
{
    return LoadStrikeRunner.WithNatsServerUrl(baseContext(), "nats://demo-cluster:4222");
}
    }
